vendor/bat-native-ads/src/bat/ads/internal/processors/contextual/text_embedding/text_embedding_html_events.cc

```diff
@@ -0,0 +1,60 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ads/internal/processors/contextual/text_embedding/text_embedding_html_events.h"
+
+#include <string>
+#include <vector>
+
+#include "base/time/time.h"
+#include "bat/ads/internal/ads_client_helper.h"
+#include "bat/ads/internal/base/instance_id_util.h"
+#include "bat/ads/internal/base/logging_util.h"
+#include "bat/ads/internal/processors/contextual/text_embedding/text_embedding_html_event_info.h"
+#include "bat/ads/internal/processors/contextual/text_embedding/text_embedding_html_events_database_table.h"
+
+namespace ads {
+
+void LogTextEmbeddingHTMLEvent(const std::string embedding_formatted,
+                               const std::string hashed_key,
+                               TextEmbeddingHTMLEventCallback callback) {
+  TextEmbeddingEventInfo text_embedding_event_info;
+  text_embedding_event_info.timestamp = base::Time::Now();
+  text_embedding_event_info.version = "";
+  text_embedding_event_info.locale = "";
+  text_embedding_event_info.hashed_key = hashed_key;
+  text_embedding_event_info.embedding = embedding_formatted;
+
+  database::table::TextEmbeddingHTMLEvents database_table;
+  database_table.LogEvent(
+      text_embedding_event_info,
+      [callback](const bool success) { callback(success); });
+}
+
+void PurgeStaleTextEmbeddingHTMLEvents(
+    TextEmbeddingHTMLEventCallback callback) {
+  database::table::TextEmbeddingHTMLEvents database_table;
+  database_table.PurgeStale(
+      [callback](const bool success) { callback(success); });
+}
+
+void GetTextEmbeddingEventsFromDatabase() {
+  database::table::TextEmbeddingHTMLEvents database_table;
+  database_table.GetAll(
+      [=](const bool success,
+          const TextEmbeddingHTMLEventList& text_embedding_html_events) {
+        if (!success) {
+          BLOG(1, "Failed to get ad events");
+          return;
+        }
+
+        for (const auto& text_embedding_html_event :
+             text_embedding_html_events) {
+          BLOG(1, "Stored embedding: " << text_embedding_html_event.embedding);
```

Comment:
Do you have an example of the `embedding` string so we can make sure nothing sensitive is logged?

Reply:
Yes here is an example of an embedding:

`0.0022974000312387943 -0.1050800010561943 0.11924000084400177 0.009181399829685688 -0.008751600049436092 -0.07865499705076218 0.08621499687433243 -0.1822900027036667 0.11021000146865845 0.012335999868810177 0.04795699939131737 -0.08016599714756012 0.020371999591588974 -0.01039700023829937 -0.011076999828219414 0.04834099858999252 0.2226399928331375 -0.004063500091433525 0.09237600117921829 0.1310099959373474 0.057036999613046646 -0.002895199926570058 0.093640998005867 0.1035500019788742 0.020274000242352486 -0.016347000375390053 -0.008080899715423584 0.004974599927663803 0.047189000993967056 -0.02022700011730194 0.03395500034093857 0.00014290999388322234 -0.15466000139713287 -0.1407500058412552 0.06276799738407135 0.04940300062298775 0.02528199926018715 -0.05231799930334091 0.025691000744700432 -0.0029416000470519066 -0.0177839994430542 -0.2582300007343292 0.031231999397277832 -0.02175300009548664 -0.007296800147742033 -0.009114399552345276 -0.02124599926173687 0.003121200017631054 -0.026854999363422394 0.018640000373125076 0.03229200094938278 -0.03679800033569336 0.1282999962568283 0.0288270004093647 -0.11404000222682953 0.002654300071299076 0.04013599827885628 -0.012041999958455563 -0.06240899860858917 -0.05087000131607056 0.0558059997856617 0.058816999197006226 0.11819999665021896 0.14539000391960144 -0.02879299968481064 0.01865600049495697 -0.0007674900116398931 -0.009551400318741798 0.11407999694347382 0.044443998485803604 -0.014248999767005444 -0.04467900097370148 -0.08436799794435501 0.011212999932467937 -0.032719001173973083 -0.0872960016131401 -0.006344399880617857 -0.041708000004291534 -0.027574999257922173 0.05158500000834465 -0.01937899924814701 0.0669580027461052 -0.044192999601364136 0.08749499917030334 -0.0768589973449707 0.04841599985957146 0.10073000192642212 -0.004007999785244465 0.10958000272512436 0.1360899955034256 0.07142800092697144 0.033854998648166656 -0.18231000006198883 -0.04600000008940697 0.03593499958515167 0.07182399928569794 -0.023062000051140785 0.027754999697208405 0.06917600333690643 -0.04210999980568886 0.006600900087505579 -0.08200199902057648 0.04217800125479698 -0.1421400010585785 -0.06413999944925308 -0.10469000041484833 0.07459700107574463 0.08612599968910217 0.031119000166654587 -0.0009229599963873625 -0.06933300197124481 0.10335999727249146 0.07001499831676483 -0.038001999258995056 0.04263899847865105 0.10802000015974045 0.011022999882698059 -0.04314599931240082 -0.08017200231552124 0.08485700190067291 0.021276000887155533 0.028064999729394913 0.01979999989271164 0.00664159981533885 0.05305900052189827 0.057436998933553696 -0.062449999153614044 -0.09670799970626831 -0.04258999973535538 0.32826000452041626 -0.03871000185608864 0.0367950014770031 0.012496000155806541 0.06176299974322319 0.020020000636577606 -0.009467099793255329 0.1011900007724762 0.00972750037908554 0.09505099803209305 -0.09535299986600876 0.0791729986667633 -0.11808999627828598 -0.08070600032806396 0.005604399833828211 -0.02033100090920925 -0.024013999849557877 -0.03942300006747246 -0.031853001564741135 -0.04758799821138382 -0.09150800108909607 0.07846699655056 0.052740998566150665 0.05014599859714508 -0.06709499657154083 0.1124500036239624 0.1642799973487854 -0.03854599967598915 -0.08369799703359604 0.018411999568343163 -0.053777001798152924 0.0939909964799881 0.06235700100660324 -0.1495400071144104 0.07167600095272064 0.015780000016093254 -0.02078999951481819 0.08054599910974503 -0.00679749995470047 -0.007619500160217285 -0.024883000180125237 -0.03188600018620491 -0.055337999016046524 0.08116800338029861 -0.014758000150322914 -0.09486699849367142 0.08221299946308136 -0.044109001755714417 0.13037000596523285 -0.08152899891138077 -0.09020800143480301 -0.03677700087428093 -0.059390999376773834 -0.01875700056552887 -0.05006900057196617 -0.029851000756025314 0.059679001569747925 0.11264999955892563 0.19935999810695648 0.20396000146865845 0.11114999651908875 -0.10401999950408936 0.19753000140190125 -0.05926999822258949 -0.020419999957084656 -0.026280999183654785 -0.034262001514434814 0.10339999943971634 0.033987998962402344 -0.11114999651908875 -0.029037000611424446 -0.2910799980163574 -0.0020316001027822495 -0.007269300054758787 -0.05777699872851372 -0.004535499960184097 0.025425000116229057 0.1517300009727478 -0.06405699998140335 0.0374940000474453 -0.0020834999158978462 0.20056000351905823 0.052257001399993896 -0.055684998631477356 0.03393099829554558 -0.094760000705719 0.08885399997234344 0.08694600313901901 -0.08627299964427948 -0.07344400137662888 -0.12352000176906586 0.05997899919748306 -0.11039000004529953 -0.053807999938726425 0.16033999621868134 0.011013999581336975 -0.026298999786376953 0.006932300049811602 0.08325999975204468 -0.011153999716043472 -0.008945699781179428 0.005445700138807297 -0.1300099939107895 -0.03999999910593033 -0.062428999692201614 -0.007527099922299385 -0.07842300087213516 -0.0565049983561039 -0.033459000289440155 -0.030987000092864037 0.1503400057554245 0.08551900088787079 -0.015785999596118927 0.028629999607801437 -0.015340999700129032 0.045667000114917755 -0.0016925999661907554 -0.01866300031542778 -0.04308899864554405 -0.035943999886512756 0.0694039985537529 0.09621799737215042 -0.0826990008354187 -0.01434599980711937 0.037338998168706894 0.1040399968624115 0.053279999643564224 0.060026999562978745 0.11275999993085861 -0.07297500222921371 0.04778999835252762 -0.05977100133895874 -0.04080900177359581 -0.05776499956846237 -0.01698100008070469 -0.053640998899936676 0.01607399992644787 0.148389995098114 0.03383500128984451 -0.0031906000804156065 -0.025103000923991203 0.03439699858427048 0.10926000028848648 -0.19649000465869904 -0.03978100046515465 0.019398000091314316 -0.0322129987180233 -0.0743890032172203 0.05730000138282776 -0.05046899989247322 -0.07716400176286697 0.16739000380039215 0.008248300291597843 -0.09762600064277649 -0.03825100138783455 -0.010374999605119228 0.07132700085639954 -0.0561939999461174 0.01995600014925003 0.05282299965620041 0.06871499866247177 -0.030354000627994537 -0.03042599931359291 0.11494000256061554 -0.039817001670598984 -0.09639400243759155 -0.0608379989862442 -0.06392499804496765 -0.05518300086259842 -0.0417729988694191 -0.0248429998755455`

Reply:
Other fields could look like:

id: `5`
created_at: `1658733634.74542`
hashed_key: `jx90rfZYZMhtPUceqMqeMp1Cgkie3BVsmWBCZAkHdL8=`

Reply:
Do we need to log this, as very verbose?

Reply:
I believe this is addressed by the BLOG level 3 comment above? I'd like to have it visible at some verbosity, otherwise I'd have to go into sqlite to check the embeddings

Reply:
Levels 0 to 6 are logged to the rewards diagnostic log. As this is only for developers level 9 is adequate as we can then debug with `--vmodule="*/bat-native-ads/*"=9`

Reply:
Changed to level 9